toolkit/tools/pkg/imagecustomizerlib/shrinkfilesystems.go

```diff
@@ -0,0 +1,152 @@
+			}
+
+			// Find the new partition end value
+			end, err := getNewPartitionEndInSectors(stdout, matchStarts[partitionNum-1][1], imageLoopDevice)
+			if err != nil {
+				return err
+			}
+
+			// Resize the partition with parted resizepart
+			_, stderr, err = shell.ExecuteWithStdin("yes" /*stdin*/, "sudo", "parted", "---pretend-input-tty", imageLoopDevice, "resizepart", strconv.Itoa(partitionNum), end)
+			if err != nil {
+				return fmt.Errorf("failed to resizepart %s with parted:\n%v", partitionLoopDevice, stderr)
+			}
+
+			// Re-read the partition table
+			err = shell.ExecuteLive(true, "flock", "--timeout", "5", imageLoopDevice, "partprobe", "-s", imageLoopDevice)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Get the start sectors of all partitions.
+func getStartSectors(imageLoopDevice string, partitionCount int) (matchStarts [][]string, err error) {
+	stdout, stderr, err := shell.Execute("sudo", "fdisk", "-l", imageLoopDevice)
+	if err != nil {
+		return nil, fmt.Errorf("fdisk failed to list partitions: \n%v", stderr)
+	}
+
+	// Since regexp.QuoteMeta doesn't escape forward slashes, handling it here
+	escapeForwardSlashesLoopDevice := strings.ReplaceAll(imageLoopDevice, "/", `\/`)
+
+	// Example line from fdisk -l output: "/dev/loop41p2   18432  103064   84633 41.3M Linux filesystem"
+	reStarts := regexp.MustCompile(`(?m:^` + escapeForwardSlashesLoopDevice + `p\d+ *(\d+).*?)`)
+	matchStarts = reStarts.FindAllStringSubmatch(stdout, -1)
+	if len(matchStarts) < partitionCount {
+		return nil, fmt.Errorf("could not find all partition starts")
+	}
+
+	return matchStarts, nil
+}
+
+// Get the filesystem size in sectors.
+func getFilesystemSizeInSectors(resize2fsOutput string, imageLoopDevice string) (filesystemSizeInSectors int, err error) {
+	// Example resize2fs output first line: "Resizing the filesystem on /dev/loop44p2 to 21015 (4k) blocks."
+	re := regexp.MustCompile(`.*to (\d+) \((\d+)([a-zA-Z])\)`)
+	// Get the block count and block size
+	match := re.FindStringSubmatch(resize2fsOutput)
+	if len(match) < 4 {
+		return 0, fmt.Errorf("failed to parse output of resize2fs")
+	}
+
+	blockCount, err := strconv.Atoi(match[1]) // Example: 10579
+	if err != nil {
+		return 0, err
+	}
+	multiplier, err := strconv.Atoi(match[2]) // Example: 4
+	if err != nil {
+		return 0, err
+	}
+	unit := match[3] // Example: 'k'
+
+	// Calculate block size
+	var blockSize int
+	const KB = 1024 // kilobyte in bytes
+	switch unit {
+	case "k":
+		blockSize = multiplier * KB
+	default:
+		return 0, fmt.Errorf("unrecognized unit %s", unit)
+	}
+
+	filesystemSizeInBytes := blockCount * blockSize
+
+	// Get the sector size
+	logicalSectorSize, _, err := diskutils.GetSectorSize(imageLoopDevice)
+	if err != nil {
+		fmt.Errorf("failed to get sector size")
+	}
+	sectorSizeInBytes := int(logicalSectorSize) // cast from uint64 to int
+
+	return int(math.Ceil(float64(filesystemSizeInBytes) / float64(sectorSizeInBytes))), nil
```

Comment:
nit:

```golang
filesystemSizeInSectors := filesystemSizeInBytes / sectorSizeInBytes
if filesystemSizeInBytes % sectorSizeInBytes != 0 {
    filesystemSizeInSectors++
}

return filesystemSizeInSectors, nil
```

This will be a lot faster than using float and won't be subject to floating point rounding errors.